lib/puppet/pops/types/p_sensitive_type.rb

```diff
@@ -0,0 +1,68 @@
+module Puppet::Pops
+module Types
+
+# A Puppet Language type that wraps sensitive information. The sensitive type is parameterized by
+# the wrapped value type.
+#
+#
+# @api public
+class PSensitiveType < Puppet::Pops::Types::PTypeWithContainedType
+
+  class Sensitive
+    def initialize(value)
+      @value = value
+    end
+
+    def unwrap
+      @value
+    end
+
+    def to_s
+      "Sensitive [value redacted]"
+    end
+
+    def inspect
+      "#<#{to_s}>"
+    end
+  end
+
+  def initialize(type = nil)
+    @type = type.nil? ? PAnyType.new : type.generalize
+  end
+
+  def instance?(o, guard = nil)
+    o.is_a?(Sensitive) && @type.instance?(o.unwrap, guard)
+  end
+
+  def self.new_function(_, loader)
+    @new_function ||= Puppet::Functions.create_loaded_function(:new_Sensitive, loader) do
+
+      dispatch :from_sensitive do
+        param 'Sensitive', :value
+      end
+
+      dispatch :from_any do
+        param 'Any', :value
+      end
+
+      def from_any(value)
+        Sensitive.new(value)
+      end
+
+      def from_sensitive(value)
+        Sensitive.new(value.unwrap)
```

Comment:
Do you want a copy? The value is immutable so could return the very same instance.